src/java.base/share/classes/jdk/internal/foreign/abi/Specializer.java

```diff
@@ -0,0 +1,854 @@
+
+                assert RETURN_BUFFER_IDX != -1;
+                emitLoad(BasicType.L, RETURN_BUFFER_IDX);
+                Class<?> valueLayoutType = emitLoadLayoutConstant(storeType);
+                emitConst(retBufOffset);
+                emitLoad(basicStoreType, valueIdx);
+                String descriptor = methodType(void.class, valueLayoutType, long.class, storeType).descriptorString();
+                mv.visitMethodInsn(INVOKEINTERFACE, MEMORY_SEGMENT_INTRN, SET_NAME, descriptor, true);
+                retBufOffset += abi.arch.typeSize(vmStore.storage().type());
+            }
+        }
+    }
+    private void emitVMLoad(Binding.VMLoad vmLoad) {
+        Class<?> loadType = vmLoad.type();
+
+        if (callingSequence.forDowncall()) {
+            // processing return
+            if (!callingSequence.needsReturnBuffer()) {
+                emitRestoreReturnValue(loadType);
+            } else {
+                assert RETURN_BUFFER_IDX != -1;
+                emitLoad(BasicType.L, RETURN_BUFFER_IDX);
+                Class<?> valueLayoutType = emitLoadLayoutConstant(loadType);
+                emitConst(retBufOffset);
+                String descriptor = methodType(loadType, valueLayoutType, long.class).descriptorString();
+                mv.visitMethodInsn(INVOKEINTERFACE, MEMORY_SEGMENT_INTRN, GET_NAME, descriptor, true);
+                retBufOffset += abi.arch.typeSize(vmLoad.storage().type());
+                typeStack.push(loadType);
+            }
+        } else {
+            // processing arg
+            emitGetInput();
+        }
+    }
+
+    private void emitDupBinding() {
+        Class<?> dupType = typeStack.peek();
+        emitDup(BasicType.of(dupType));
+        typeStack.push(dupType);
+    }
+
+    private void emitUnboxAddress() {
+        assert Addressable.class.isAssignableFrom(typeStack.pop());
+        mv.visitMethodInsn(INVOKEINTERFACE, ADDRESSABLE_INTRN, ADDRESS_NAME, ADDRESS_DESC, true);
+        emitToRawLongValue();
+        typeStack.push(long.class);
+    }
+
+    private void emitBufferLoad(Binding.BufferLoad bufferLoad) {
+        Class<?> loadType = bufferLoad.type();
+        long offset = bufferLoad.offset();
+
+        assert typeStack.pop() == MemorySegment.class;
+
+        Class<?> valueLayoutType = emitLoadLayoutConstant(loadType);
+        emitConst(offset);
+        String descriptor = methodType(loadType, valueLayoutType, long.class).descriptorString();
+        mv.visitMethodInsn(INVOKEINTERFACE, MEMORY_SEGMENT_INTRN, GET_NAME, descriptor, true);
+        typeStack.push(loadType);
+    }
+
+    private void emitCopyBuffer(Binding.Copy copy) {
+        long size = copy.size();
+        long alignment = copy.alignment();
+
+        assert typeStack.pop() == MemorySegment.class;
+
+        // operand/srcSegment is on the stack
+        // copy(MemorySegment srcSegment, long srcOffset, MemorySegment dstSegment, long dstOffset, long bytes)
```

Comment:
I found those comments confusing, as they almost look like commented out code :-)

Reply:
I tried to make the comments a bit more clear. It's intended as a mental note of the signature of the method that is being invoked in the bytecode below.